src/RefinedAst.hs

```diff
@@ -63,71 +69,78 @@ data MType
 --  | Mapping (Map MType MType)
   deriving (Eq, Ord, Show, Read)
 
---data T_List t
---data T_Tuple
-
 data StorageUpdate
-  = IntUpdate (TStorageItem Int) (Exp Int)
+  = IntUpdate (TStorageItem Integer) (Exp Integer)
   | BoolUpdate (TStorageItem Bool) (Exp Bool)
   | BytesUpdate (TStorageItem ByteString) (Exp ByteString)
   deriving (Show)
 
 data StorageLocation
-  = IntLoc (TStorageItem Int)
+  = IntLoc (TStorageItem Integer)
   | BoolLoc (TStorageItem Bool)
   | BytesLoc (TStorageItem ByteString)
-  deriving (Show)
-
+  deriving (Show, Eq)
 
 data TStorageItem a where
-  DirectInt    :: Id -> TStorageItem Int
-  DirectBool   :: Id -> TStorageItem Bool
-  DirectBytes  :: Id -> TStorageItem ByteString
-  MappedInt    :: Id -> NonEmpty ReturnExp -> TStorageItem Int
-  MappedBool   :: Id -> NonEmpty ReturnExp -> TStorageItem Bool
-  MappedBytes  :: Id -> NonEmpty ReturnExp -> TStorageItem ByteString
+  DirectInt    :: Id -> Id -> TStorageItem Integer
+  DirectBool   :: Id -> Id -> TStorageItem Bool
+  DirectBytes  :: Id -> Id -> TStorageItem ByteString
+  MappedInt    :: Id -> Id -> NonEmpty ReturnExp -> TStorageItem Integer
+  MappedBool   :: Id -> Id -> NonEmpty ReturnExp -> TStorageItem Bool
+  MappedBytes  :: Id -> Id -> NonEmpty ReturnExp -> TStorageItem ByteString
 
 deriving instance Show (TStorageItem a)
+
+instance Eq (TStorageItem a) where
```

Comment:
can this be derived?
```
deriving instance Eq (TStorageItem a)
```

Reply:
oh, yes it can, dunno how I missed that. Thanks!!!